File: 0599-minimum-index-sum-of-two-lists/0599-minimum-index-sum-of-two-lists.cpp
class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        int i=0, count = 0,min = INT_MAX;
        vector<string> *ans = new vector<string>();
        unordered_map<string, int> map;
        vector<string> m = list1.size() < list2.size() ? list1 : list2;
        vector<string> list = list1.size() >= list2.size() ? list1 : list2;
        for(string s:m) {
            if(map.count(s) == 0)
                map[s]=i;
            i++;
        }
        i=0;
        for(string s:list) {
            if(map.count(s) == 1) {
                count = i + map[s];
                if(count < min) {
                    min = count;
                    ans = new vector<string>();
                }
                if(count == min)
                    ans->push_back(s);
            }
            i++;
        }
        return *ans;
    }
};

File: 0599-minimum-index-sum-of-two-lists/0599-minimum-index-sum-of-two-lists.java
class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        int i=0, count = 0,min = Integer.MAX_VALUE;
        List<String> ans = new ArrayList<>();
        Map<String, Integer> map = new HashMap<>();
        String[] m = list1.length < list2.length ? list1 : list2;
        String[] list = list1.length >= list2.length ? list1 : list2;
        for(String s:m) {
            map.putIfAbsent(s,i);
            i++;
        }
        i=0;
        for(String s:list) {
            if(map.containsKey(s)) {
                count = i + map.get(s);
                if(count < min) {
                    min = count;
                    ans = new ArrayList<>();
                }
                if(count == min)
                    ans.add(s);
            }
            i++;
        }
        return ans.toArray(new String[0]);
    }
}